Todo/AppDelegate.swift

```diff
@@ -0,0 +1,420 @@
+            NSLog("Cannot close database: %@", error)
+        }
+        let oldUsername = Session.username
+        Session.username = nil
+        login(oldUsername)
+    }
+
+    func showApp() {
+        guard let root = window?.rootViewController, storyboard = root.storyboard else {
+            return
+        }
+
+        let controller = storyboard.instantiateInitialViewController()
+        window!.rootViewController = controller
+    }
+
+    // MARK: - LoginViewControllerDelegate
+
+    func login(controller: UIViewController, withUsername username: String,
+        andPassword password: String) {
+            processLogin(controller, withUsername: username, withPassword: password)
+    }
+
+    func processLogin(controller: UIViewController, withUsername username: String,
+        withPassword password: String, withNewPassword newPassword: String? = nil) {
+            do {
+                try startSession(withUsername: username, withPassword: password,
+                    withNewPassword: newPassword)
+            } catch let error as NSError {
+                if error.code == 401 {
+                    handleEncryptionError(controller, withUsername: username,
+                        withPassword: password)
+                } else {
+                    Ui.showMessageDialog(
+                        onController: controller,
+                        withTitle: "Error",
+                        withMessage: "Login has an error occurred, code = \(error.code).")
+                    NSLog("Cannot start a session: %@", error)
+                }
+            }
+    }
+
+    func handleEncryptionError(controller: UIViewController, withUsername username: String,
+        withPassword password: String) {
+            Ui.showTextInputDialog(
+                onController: controller,
+                withTitle: "Password Changed",
+                withMessage: "Please enter your old password to migrate your database.",
+                withTextFieldConfig: { (textField) -> Void in
+                    textField.placeholder = "old password"
+                    textField.secureTextEntry = true
+                    textField.autocapitalizationType = .None
+                },
+                withOkConfig: { () -> (title: String, style: UIAlertActionStyle) in
+                    return ("Migrate", UIAlertActionStyle.Default)
+                },
+                withCancelConfig: { () -> (title: String, style: UIAlertActionStyle) in
+                    return ("Delete", UIAlertActionStyle.Destructive)
+                },
+                onOk: { (oldPassword) -> Void in
+                    self.processLogin(controller, withUsername: username,
+                        withPassword: oldPassword, withNewPassword: password)},
+                onCancel: { () -> Void in
+                    // Force delete the database:
+                    let dir = NSURL(fileURLWithPath: CBLManager.sharedInstance().directory)
```

Comment:
Even though it's necessary, this is a hack. It would be better to move it out to a separate function, with a comment explaining why it's necessary.